database_sanitizer/utils/postgres.py

```diff
@@ -6,140 +6,114 @@
 Documentation about copy command and the text format used by it can be found
 from:
 https://www.postgresql.org/docs/9.2/static/sql-copy.html
+
+For decoding we use a regular expression to find the escape sequences
+and invoke `unescape_single_character` function for each occurence.
+Allowed escape sequences are precalculated into `DECODE_MAP` to make the
+lookups faster.
+
+For encoding we use a string translation table `ENCODE_TRANSLATE_TABLE`,
+which maps the "forbidden" characters to escape sequences.  This is used
+with `str.translate`, which is very fast way to escape characters.
 """
 
 from __future__ import unicode_literals
 
-import six
+import itertools
+import re
 import string
 
+import six
 
 #: Representation of NULL value in Postgres COPY statement.
 POSTGRES_COPY_NULL_VALUE = "\\N"
 
+ENCODE_MAP = {
+    '\\': '\\\\',
+    '\b': '\\b',
+    '\f': '\\f',
+    '\n': '\\n',
+    '\r': '\\r',
+    '\t': '\\t',
+    '\v': '\\v',
+}
+
+ENCODE_TRANSLATE_TABLE = [
+    ENCODE_MAP.get(six.unichr(n), six.unichr(n))
+    for n in range(256)
+]
+
+DECODE_REGEX = re.compile(r"""
+\\                 # a backslash
+(?:                # followed by one of these (in non-capturing parenthesis):
+    [0-7]{1,3}         # 1, 2 or 3 octal digits
+    |                  # or
+    x[0-9a-fA-F]{1,2}  # 'x' followed by 1 or 2 hexadecimal digits
+    |                  # or
+    .                  # any character
+    |                  # or
+    \Z                 # end of string
+)
+""", re.VERBOSE)
 
-def decode_copy_value(value):
-    """
-    Decodes value received as part of Postgres `COPY` command.
-
-    :param value: Value to decode.
-    :type value: str
 
-    :return: Either None if the value is NULL string, or the given value where
-             escape sequences have been decoded from.
-    :rtype: str|None
-    """
+def decode_copy_value(value):
     # Test for null values first.
     if value == POSTGRES_COPY_NULL_VALUE:
         return None
 
-    index = 0
-    length = len(value)
-    result = None
+    # If there is no backslash present, there's nothing to decode.
+    #
+    # This early return provides a little speed-up, because it's very
+    # common to not have anything to decode and then simple search for
+    # backslash is faster than the regex sub below.
+    if '\\' not in value:
+        return value
 
-    while index < length:
-        c = value[index]
+    return DECODE_REGEX.sub(unescape_single_character, value)
 
-        if c != "\\":
-            if result is not None:
-                result += c
-            index += 1
-            continue
 
-        if result is None:
-            result = value[:index]
-        index += 1
-        if index >= length:
+def unescape_single_character(match):
```

Comment:
Docstrings, to keep things consistent with the rest of the code 😄

Reply:
It has now.

Also added docstrings to couple other places.